pkg/block/s3/retyer.go

```diff
@@ -0,0 +1,41 @@
+package s3
+
+import (
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws/client"
+	"github.com/aws/aws-sdk-go/aws/request"
+)
+
+// errConnectionResetRetryer implements the `request.Retryer` interface and allows for
+// customization of the retry behaviour of an AWS client.
+// errConnectionResetRetryer was taken from cockroachdb
+// https://github.com/cockroachdb/cockroach/blob/7a8cac9b11036dcf401453eb36e531fe381b70b4/pkg/cloud/amazon/s3_storage.go#L95
+type errConnectionResetRetryer struct {
+	client.DefaultRetryer
+}
+
+// isErrReadConnectionReset returns true if the underlying error is a read
+// connection reset error.
+//
+// NB: A read connection reset error is thrown when the SDK is unable to read
+// the response of an underlying API request due to a connection reset. The
+// DefaultRetryer in the AWS SDK does not treat this error as a retryable error
+// since the SDK does not have knowledge about the idempotence of the request,
+// and whether it is safe to retry -
+// https://github.com/aws/aws-sdk-go/pull/2926#issuecomment-553637658.
+//
+// In lakeFS all operations with s3 (read, write, list) are considered idempotent,
+// and so we can treat the read connection reset error as retryable too.
+func isErrReadConnectionReset(err error) bool {
+	// The error string must match the one in
+	// github.com/aws/aws-sdk-go/aws/request/connection_reset_error.go. This is
+	// unfortunate but the only solution until the SDK exposes a specialized error
+	// code or type for this class of errors.
+	return err != nil && strings.Contains(err.Error(), "read: connection reset")
```

Comment:
This condition is not only strange, it also looks like the [_opposite_ of `isConnectionReset`](https://github.com/aws/aws-sdk-go/blob/31e1e85e8c0473ce97aa60099e5bd491c09f19da/aws/request/connection_reset_error.go#L8-L10).  For some reason, AWS consider a "connection reset" to be a connectionReset only if it is **not** a "read: connection reset".

I don't understand it.

Reply:
Yap - I think the linked code is confusing. The appropriate function name should be `isErrConnectionResetAndShouldRetry`

Reply:
Can you do something like what this [StackOverflow answer](https://stackoverflow.com/a/64146772/192263) suggests?  It should work _if_ AWS errors wrap OS errors (in the golang sense) -- and if it is good, it is much better than looking at the error message.

Also worried about the "read:" prefix.

Reply:
Had a look at AWSErr, and it seems _not_ to wrap.  But you can try to cast to an AWSError, and if that succeeds call [`OrigErr`](https://github.com/aws/aws-sdk-go/blob/31e1e85e8c0473ce97aa60099e5bd491c09f19da/aws/awserr/types.go#L90-L93).  I don't see OrigErr as returning the BatchError case except for errors from  the S3 manager, so probably safe to ignore that.

So something like
```go
if awsErr, ok := err.(awserr.Error) {
	err = awsErr.OrigErr()
}
return errors.Is(err, syscall.ECONNRESET)
```
might work!!?!
}